### PR TITLE
Add banner bean

### PR DESCRIPTION
**Pull request by michael-simons**
If the banner hasn't been turned off and was printed to either the log or the console, add it as a singleton to the application context.
The name of the bean can be configured, which i have documented in the docs.
This way the feature can also be turned off.

I've changed changed the return value of SpringApplicationBannerPrinter#print, but that shouldn't cause problems as the class is only package level visible. I've also added a new parameter to SpringApplication#prepareContext, but this method is private.

- [x] I have signed the CLA

### CODE REVIEW

File: spring-boot-docs/src/main/asciidoc/spring-boot-features.adoc [hunk 1]
```diff
@@ -94,6 +94,9 @@ You can also use the `spring.main.banner-mode` property to determine if the bann
 to be printed on `System.out` (`console`), using the configured logger (`log`) or not
 at all (`off`).
 
+The printed banner will be registered as a singleton bean under the name `springBanner`.
+This can be turned of by setting `banner.bean.name` to a blank value.
```
Comment (spencergibb) [inline spring-boot-docs/src/main/asciidoc/spring-boot-features.adoc]: off not of

File: spring-boot-docs/src/main/asciidoc/appendix-application-properties.adoc [hunk 2]
```diff
@@ -36,6 +36,7 @@ content into your application; rather pick only the properties that you need.
 	banner.image.hight= # Height of the banner image in chars (default based on image height)
 	banner.image.margin= # Left hand image margin in chars (default 2)
 	banner.image.invert= # If images should be inverted for dark terminal themes (default false)
+    banner.bean.name=springBanner # The of the bean under which the banner is registered, an empty property disables registration of the bean.
```
Comment (snicoll) [inline spring-boot-docs/src/main/asciidoc/appendix-application-properties.adoc]: We don't do that and I really see a compelling reason to be able to customize the bean name. If we need an early lookup of some kind, let's settle with a default name that we put in the doc instead.

File: spring-boot/src/main/java/org/springframework/boot/SpringApplication.java [hunk 3]
```diff
@@ -353,6 +361,10 @@ private void prepareContext(ConfigurableApplicationContext context,
 		// Add boot specific singleton beans
 		context.getBeanFactory().registerSingleton("springApplicationArguments",
 				applicationArguments);
+		String bannerBeanName = environment.getProperty(BANNER_BEAN_NAME_PROPERTY, String.class, "springBanner").trim();
```
Comment (snicoll) [inline spring-boot/src/main/java/org/springframework/boot/SpringApplication.java]: We should just do it by default, something like `springBootBanner` (see what was done in `LoggingApplicationListener` for the `LoggingSystem`).

File: spring-boot/src/main/java/org/springframework/boot/SpringApplication.java [hunk 4]
```diff
@@ -536,19 +548,22 @@ protected void configureProfiles(ConfigurableEnvironment environment, String[] a
...
 	 * @param environment the environment
+	 * @return The banner that was printed to the console
 	 * @see #setBannerMode
 	 */
-	protected void printBanner(Environment environment) {
+	protected Banner printBanner(Environment environment) {
 		ResourceLoader resourceLoader = this.resourceLoader != null ? this.resourceLoader
 				: new DefaultResourceLoader(getClassLoader());
-		SpringApplicationBannerPrinter banner = new SpringApplicationBannerPrinter(resourceLoader,
+		SpringApplicationBannerPrinter bannerPrinter = new SpringApplicationBannerPrinter(resourceLoader,
```
Comment (snicoll) [inline spring-boot/src/main/java/org/springframework/boot/SpringApplication.java]: Seems like a glitch, isn't it?

File: spring-boot-docs/src/main/asciidoc/appendix-application-properties.adoc (hunk 2 above)
Comment (michael-simons) [inline spring-boot-docs/src/main/asciidoc/appendix-application-properties.adoc]: Thanks, Stéphane, i didn't know, but i'm totally fine with. I'll change it.

File: spring-boot/src/main/java/org/springframework/boot/SpringApplication.java (hunk 4 above)
Comment (michael-simons) [inline spring-boot/src/main/java/org/springframework/boot/SpringApplication.java]: Actually not. It's the banner printer, not the banner.

Comment (snicoll) [inline spring-boot/src/main/java/org/springframework/boot/SpringApplication.java]: Ah, right. I didn't look carefully at the type, sorry.

**Comment by philwebb** [comment]
Thanks! I've made a small tweak to keep the old `printBanner` method signature. It's now deprecated to persuade people to jump to use the `Banner` interface.

**Pull request closed without merging**